tree/treeplayer/inc/ROOT/TDFActionHelpers.hxx

```diff
@@ -92,6 +93,30 @@ public:
    ULong64_t &PartialUpdate(unsigned int slot);
 };
 
+template<typename ProxiedVal_t>
+class ReportHelper {
+   const std::shared_ptr<TCutFlowReport> fReport;
+   std::weak_ptr<ProxiedVal_t> fProxiedWPtr;
+   bool fReturnEmptyReport;
+public:
+   using BranchTypes_t = TypeList<>;
+   ReportHelper(const std::shared_ptr<TCutFlowReport> &report, const std::shared_ptr<ProxiedVal_t> &pp, bool emptyRep)
+   :fReport(report), fProxiedWPtr(pp), fReturnEmptyReport(emptyRep) {};
+   ReportHelper(ReportHelper &&) = default;
+   ReportHelper(const ReportHelper &) = delete;
+   void InitSlot(TTreeReader *, unsigned int) {}
+   void Exec(unsigned int /* slot */ ) {}
+   void Initialize() { /* noop */ }
+   void Finalize()
+   {
+      // We need the weak_ptr in order to avoid crashes at tear down
+      if (!fReturnEmptyReport && !fProxiedWPtr.expired())
+         fProxiedWPtr.lock()->Report(*fReport);
+   }
+   TCutFlowReport &PartialUpdate(unsigned int /* slot */) {return *fReport;}; // it's a shared_ptr and thread safe
```

Comment:
If I understand correctly `fReport` is not updated during the event loop, so `OnPartialResult` would give the user an empty `TCutFlowReport`for the whole event loop.
At least until some time ago there was a way to opt out of the `OnPartialResult` technology, e.g. by not implementing this method at all (can't remember exactly at the moment). Maybe `Report` should not allow an `OnPartialResult`, at least for now.

Reply:
This is indeed awkward. This action is a bit different than the others and it is impossible, at least for now, to have partial results. It would be cool to see a table of efficiencies updated live during the event loop but somehow the cost is a bit high at the moment. I'll try to remove this method and see what happens.

Reply:
I just verified that indeed `TAction` throws an exception when `OnPartialResult` is invoked but the `PartialResult` method of the corresponding helper is not implemented

Reply:
I do homework fast.